### PR TITLE
Xbox XEX file format support

### LINKED ISSUE #10: Xbox XEX file format support
XEX is the executable file format used by the Xbox 360 operating system. It seems to be a crypto and packing container for PPC PE executable files, comparable to UPX or TEEE Burneye. It's likely the 360 knows how to grab the section it needs into memory and decrypt/decompress on demand, instead of traditional all at once extraction.

https://free60project.github.io/wiki/XEX/

[xexdump.zip](https://github.com/rizinorg/ideas/files/6129270/xexdump.zip)
[xextools-0.2.tar.bz2.zip](https://github.com/rizinorg/ideas/files/6129271/xextools-0.2.tar.bz2.zip)

### PR DESCRIPTION
**Pull request by smac89**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Implementing: https://github.com/rizinorg/ideas/issues/10 for GSoC
...

**Test plan**

Attempt to read the xex file found at https://web.archive.org/web/20070311061307/http://assets.xbox.com/en-us/backwardcompatibility/default.zip
...

**Closing issues**

closes rizinorg/ideas#10
...

### CODE REVIEW

**Comment by XVilka** [comment]
I would also start from 1) sending a PR with example file to the rizin-testbins (please be sure it's not covered by copyright protection first) 2) crafting a simple test from it.

File: librz/bin/p/bin_xex.c [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	if (rz_buf_size(b) >= XEX_MAGIC_SIZE) {
+		ut8 buff[XEX_MAGIC_SIZE];
+		rz_buf_read_at(b, XEX_MAGIC_OFFSET, buff, XEX_MAGIC_SIZE);
+		return (!memcmp(buff, XEX_MAGIC, XEX_MAGIC_SIZE));
+	}
+	return false;
+}
+
+static bool load_buffer(RzBinFile *bf, void **bin_obj, RzBuffer *b, ut64 loadaddr, Sdb *sdb) {
+	eprintf("[xex] load_buffer");
```
Comment (wargio) [inline librz/bin/p/bin_xex.c]: to be sure to have everything aligned, you can use clang-format:
```bash
cd /path/to/rizin
./sys/clang-format.py -f librz/bin/p/bin_xex.c
```
Be sure to have installed `gitpython` via `sudo pip install gitpython` and clang-format

Comment (smac89) [reply librz/bin/p/bin_xex.c]: Thanks I just have to remember to run it before committing

File: librz/bin/p/bin_xex.c [hunk 2]
```diff
@@ -12,24 +12,46 @@
  * \param b
  */
 static bool check_buffer(RzBuffer *b) {
-    eprintf("[xex] check_buffer");
+	eprintf("[xex] check_buffer\n");
 	if (rz_buf_size(b) >= XEX_MAGIC_SIZE) {
 		ut8 buff[XEX_MAGIC_SIZE];
 		rz_buf_read_at(b, XEX_MAGIC_OFFSET, buff, XEX_MAGIC_SIZE);
-		return (!memcmp(buff, XEX_MAGIC, XEX_MAGIC_SIZE));
+		return 0 == memcmp(buff, XEX_MAGIC, XEX_MAGIC_SIZE);
```
Comment (wargio) [inline librz/bin/p/bin_xex.c]: please use `!memcmp(buff, XEX_MAGIC, XEX_MAGIC_SIZE)`

**Comment by smac89** [comment]
@XVilka @wargio  I've created a simple test and it looks like it is passing. My issue now is with how the plugin should work going forward. So far it can parse the `header` along with the optional `header`.

My problem now is knowing what kind of information is required from the plugin and how to present that, and where. At this point, I'm just kinda guessing what parts of the plugin API I need to implement. I think some guidance will be helpful here. Especially understanding how rizin will use the plugin will be most helpful 

Also considering the scope of what still needs to be done and keeping in mind that GSoC applications deadline is in 3 days, what else do I need to do with this PR to present it as a micro-task with my application?

**Comment by XVilka** [comment]
@smac89 regarding loading information, take a look at some simple format like `librz/bin/format/ne/ne.c`
Apart from header the most important things are:
- code sections
- data sections
- symbols (imports, exports, etc)
- entry point (if applicable)

File: librz/bin/p/bin_xex.c [hunk 3]
```diff
@@ -49,29 +49,29 @@ static void init_header(RzBinFile *bf) {
  */
 static void destroy(RzBinFile *bf) {
 	eprintf("[xex] destroy\n");
-	xex_destroy_bin(&bf->o->bin_obj);
+	xex_destroy_bin((RzBinXex **)(&bf->o->bin_obj));
 
 	rz_return_if_fail(NULL == bf->o->bin_obj);
```
Comment (wargio) [inline librz/bin/p/bin_xex.c:54]: just use `rz_warn_if_fail(bf->o->bin_obj);`

File: librz/bin/p/bin_xex.c [hunk 4]
```diff
@@ -49,29 +49,29 @@ static void init_header(RzBinFile *bf) {
...
 	ret->big_endian = 1;
 	ret->has_crypto = 1;
-	ret->file = strdup(arch->file);
+	ret->file = strdup(bf->file);
 	ret->type = strdup("Xbox 360 XEX file");
 	ret->machine = strdup("Xbox system software");
 	ret->os = strdup("Xbox 360");
-	ret->arch = strdup("x86_64");
-	ret->bits = 8;
+	ret->arch = strdup("ppc64");
```
Comment (wargio) [inline librz/bin/p/bin_xex.c]: arch is `ppc`, not `ppc64`

Comment (smac89) [reply librz/bin/p/bin_xex.c]: You may be looking at the old PR. The most recent uses `ppc`

**Comment by XVilka** [comment]
@smac89 do you plan to continue working on this PR?

**Comment by XVilka** [comment]
I am closing this since there was no response for months and nobody continues to work on.

**Pull request closed without merging**